pkgsx86Darwin: init

Atemu opened the pull request:
Many packages don't work on aarch64-darwin yet or are x86-only binary packages.

This gives aarch64-darwin users easy access to x86_64-darwin packages to run via
Rosetta 2.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Atemu committed:
pkgsx86Darwin: init

Many packages don't work on aarch64-darwin yet or are x86-only binary packages.

This gives aarch64-darwin users easy access to x86_64-darwin packages to run via
Rosetta 2.

"x86" was used as a short-hand instead of x86_64 because x86_64-darwin is the
only x86 Darwin package set.

Atemu commented on pkgs/top-level/stage.nix:
```diff
@@ -213,6 +213,18 @@ let
     } else throw "i686 Linux package set can only be used with the x86 family.";
 
+    # x86_64-darwin packages for aarch64-darwin users to use with Rosetta for incompatible packages
+    pkgsDarwinIntel = if stdenv.hostPlatform.isDarwin then nixpkgsFun {
```
I'm not 100% sure about the name.

Other candidates I had in my head were `pkgsIntelDarwin` and `pkgsx86Darwin`.

Suggestions welcome.

Atemu commented:
I'm not 100% sure about the name.

Other candidates I had in my head were `pkgsIntelDarwin` and `pkgsx86Darwin`.

Suggestions welcome.

Atemu commented:
I was gravitating more towards the shorter `pkgsx86Darwin` since `i686-darwin` is not a thing.

Another thought crossed my mind while typing this: Why don't we generate such a set for all of our systems combinations?

Sciencentistguy commented:
I don't like `Intel` being present in the name, tbh. `pkgsx86Darwin` is probably the best option suggested so far imo

Atemu commented:
> I think in general it's very simple to override the system of the package set, so there's no need for a specific set, unless there's more complex changes to be done to make it work, or a particular reason for it to exist.

Well for one, the premise isn't true: It's not at all simple to override the system of a package set AFAIK. You need to re-import nixpkgs which non-trivial in everything but the simplest cases. Non-trivial example:

```Nix
{ pkgs, ... }:
{
  nixpkgs.overlays = [ (final: prev: { baz = ...; }) ];
  foo = with pkgs; [ bar pkgsx86Darwin.baz ];
}
```

> rosetta

This package set is that it's technically also usable under x86_64-darwin. This actually has a use-case in cross-arch configs.

In the example above, we'd get aarch64 bar and x86 baz on aarch64 while we'd get x86 versions of both on x86; it's compatible.

Atemu commented:
I may not have been clear enough but pkgsx86Darwin would not be a mixed-arch set. What I meant is that pkgsx86Darwin == pkgs when system == x86_64-darwin and you therefore don't need special handling when using the same Nix expressions for aarch64 and x86.

toonn commented on pkgs/top-level/stage.nix:
```diff
@@ -213,6 +213,18 @@ let
     } else throw "i686 Linux package set can only be used with the x86 family.";
 
+    # x86_64-darwin packages for aarch64-darwin users to use with Rosetta for incompatible packages
+    pkgsDarwinIntel = if stdenv.hostPlatform.isDarwin then nixpkgsFun {
+      overlays = [ (self': super': {
+        pkgsDarwinIntel = super';
```
```suggestion
    pkgsx86_64Darwin = if stdenv.hostPlatform.isDarwin then nixpkgsFun {
      overlays = [ (self': super': {
        pkgsx86_64Darwin = super';
```

toonn commented on pkgs/top-level/stage.nix:
```diff
@@ -213,6 +213,18 @@ let
+          cpu = lib.systems.parse.cpuTypes.x86_64;
+        };
+      };
+    } else throw "x64 Darwin package set can only be used on Darwin systems.";
```
```suggestion
    } else throw "x86_64 Darwin package set can only be used on Darwin systems.";
```

toonn reviewed (commented):
Looks ok to me. I'm partial to being explicit and using `pkgsx86_64Darwin`.

reckenrode commented:
I don’t have much to add except I agree regarding the name. Considering that macOS 10.12 (the version supported by the default x86_64-darwin SDK) supports 32-bit processes, I can see there being some confusion with the name even though nixpkgs doesn’t (and shouldn’t) support 32-bit Darwin.

One thing I wonder is how usable `pkgsCross.x86_64-darwin` should be from aarch64-darwin. I get an infinite recursion when trying to do `nix build nixpkgs#pkgsCross.x86_64-darwin.hello`, but if that were fixed, would it be preferable to `pkgsx86_64Darwin.hello` in most cases (due to using aarch64-darwin `nativeBuildInputs` where possible)?

reckenrode commented:
Hmm, no. Never mind the last bit. It wouldn’t help if `nativeBuildInput` didn’t work on aarch64-darwin.

uri-canva reviewed (changes requested):
I think something went wrong with a rebase or something, because @toonn's review comments are resolved, but they're not applied, I've resubmitted them.

Atemu committed:
pkgsx86Darwin -> pkgsx86_64Darwin

The majority of Darwin maintainers was in favour of being explicit about the bitness

Co-authored-by: Uri Baghin <uri@canva.com>
